alpaca_trade_api/stream2.py

```diff
@@ -271,23 +276,46 @@ def run(self, initial_channels=[]):
         initial_channels is the channels to start with.
         '''
         loop = self.loop
-        try:
-            loop.run_until_complete(self.subscribe(initial_channels))
-            loop.run_until_complete(self.consume())
-        except KeyboardInterrupt:
-            logging.info("Exiting on Interrupt")
-        finally:
-            loop.run_until_complete(self.close())
-            loop.close()
-
-    async def close(self):
-        '''Close any of open connections'''
+        should_renew = True  # should renew connection if it disconnects
+        while should_renew:
+            try:
+                if loop.is_closed():
+                    self.loop = asyncio.new_event_loop()
+                    loop = self.loop
+                loop.run_until_complete(self.subscribe(initial_channels))
+                loop.run_until_complete(self.consume())
+            except KeyboardInterrupt:
+                logging.info("Exiting on Interrupt")
+                should_renew = False
+            except Exception as e:
+                logging.error(f"error while consuming ws messages: {e}")
+                loop.run_until_complete(self.close(should_renew))
+                if loop.is_running():
+                    loop.close()
+
+    async def close(self, renew):
+        """
+        Close any of open connections
+        :param renew: should re-open connection?
+        """
         if self.trading_ws is not None:
             await self.trading_ws.close()
             self.trading_ws = None
         if self.data_ws is not None:
             await self.data_ws.close()
             self.data_ws = None
+        if renew:
```

Comment:
When this renews, it doesn't seem to retain the channels previously handled. After this renewal, the consumer just hangs. Is there an example of using this stream connection with this new code?